hphp/runtime/vm/jit/vasm-flags.cpp

```diff
@@ -0,0 +1,215 @@
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "hphp/runtime/vm/jit/vasm.h"
+
+#include "hphp/runtime/vm/jit/abi.h"
+#include "hphp/runtime/vm/jit/containers.h"
+#include "hphp/runtime/vm/jit/vasm-gen.h"
+#include "hphp/runtime/vm/jit/vasm-instr.h"
+#include "hphp/runtime/vm/jit/vasm-print.h"
+#include "hphp/runtime/vm/jit/vasm-unit.h"
+#include "hphp/runtime/vm/jit/vasm-util.h"
+#include "hphp/runtime/vm/jit/vasm-visit.h"
+
+#include "hphp/util/arch.h"
+
+TRACE_SET_MOD(vasm);
+
+namespace HPHP { namespace jit {
+
+namespace {
+
+///////////////////////////////////////////////////////////////////////////////
+
+struct Env {
+  Vunit& unit;
+
+  struct DefUses {
+    Vinstr*               def;  // The single def of a VregSF
+    jit::vector<Vinstr*>  uses; // The uses of a VregSF
+  };
+
+  // Map of VregSF to its def and uses.
+  jit::hash_map<size_t, DefUses> sf_def_uses;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+
+Vflags getFlags(const Vinstr* inst) {
+  switch (inst->op) {
+#define O(name, imms, uses, defs) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+    }
+#define Inone
+#define IF(fl) return i.fl;
+#define IC(cc)
+#define I(x)
+    VASM_OPCODES
+#undef I
+#undef IC
+#undef IF
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+void setFlags(Vinstr* inst, Vflags flags) {
+  switch (inst->op) {
+#define O(name, imms, uses, defs) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+    }
+#define Inone
+#define IF(fl) i.fl = flags; return;
+#define IC(cc)
+#define I(x)
+    VASM_OPCODES
+#undef I
+#undef IC
+#undef IF
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/*
+ * For each use of an SF, return the StatusFlags bits needed by the use's CC.
+ *
+ * Specializations are below, which can additionally change the CC if the
+ * platform doesn't have a simple CC mapping for the given def and use.
+ */
+template<typename Inst>
+Vflags annotateSFUses(Env& env, Inst& inst) {
+  return static_cast<Vflags>(required_flags(inst.cc));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+Vflags annotateSFUses(Env& env, jcc& inst) {
+  if (env.sf_def_uses[inst.sf].def->op == Vinstr::ucomisd) {
+    switch (inst.cc) {
+    case CC_AE: // CC_NB
+      inst.cc = CC_GE;
+      break;
+    default:
+      break;
+    }
+  }
+  return static_cast<Vflags>(required_flags(inst.cc));
```

Comment:
Would be nice to structure this so that the specializations don't have to repeat the boilerplate `return static_cast<Vflags>(required_flags(inst.cc));`